api/client.go

```diff
@@ -0,0 +1,153 @@
+	case loadtest.UserControllerSimple:
+		var scc *simplecontroller.Config
+		scc, ok := ucConfig.(*simplecontroller.Config)
+		if !ok {
+			return nil, errors.New("agent: ucConfig has the wrong type")
+		}
+		data.SimpleControllerConfig = scc
+	case loadtest.UserControllerSimulative:
+		scc, ok := ucConfig.(*simulcontroller.Config)
+		if !ok {
+			return nil, errors.New("agent: ucConfig has the wrong type")
+		}
+		data.SimulControllerConfig = scc
+	default:
+		return nil, errors.New("agent: UserController type is not set")
+	}
+
+	configData, err := json.Marshal(data)
+	if err != nil {
+		return nil, err
+	}
+
+	url := c.serverURL + agentApiSuffix + "create?id=" + id
+	resp, err := c.httpClient.Post(url, "application/json", bytes.NewReader(configData))
+	if err != nil {
+		return nil, fmt.Errorf("agent: post request failed: %w", err)
+	}
+	defer resp.Body.Close()
+
+	var res agentResponse
+	err = json.NewDecoder(resp.Body).Decode(&res)
+	if err != nil {
+		return nil, fmt.Errorf("agent: failed to decode load-test agent api response: %w", err)
+	}
+	if res.Error != "" {
+		return nil, fmt.Errorf("agent: load-test agent api request error: %s", res.Error)
+	} else if resp.StatusCode != http.StatusCreated {
+		return nil, fmt.Errorf("agent: bad response status code %d", resp.StatusCode)
+	}
+
+	return &Agent{
+		id:     res.Id,
+		apiURL: c.serverURL + agentApiSuffix + res.Id,
+		client: c,
+	}, nil
+}
+
+// CreateCoordinator creates a new coordinator resource with the given id and configs.
+// It returns an initialized instance of Coordinator or an error in case of failure.
+func (c *Client) CreateCoordinator(id string, coordConfig *coordinator.Config, ltConfig *loadtest.Config) (*Coordinator, error) {
+	if coordConfig == nil {
+		return nil, errors.New("coordinator: coordConfig should not be nil")
+	}
+	if ltConfig == nil {
+		return nil, errors.New("coordinator: ltConfig should not be nil")
+	}
+
+	data := struct {
+		CoordinatorConfig *coordinator.Config
+		LoadTestConfig    *loadtest.Config
+	}{
+		coordConfig,
+		ltConfig,
+	}
+
+	configData, err := json.Marshal(data)
+	if err != nil {
+		return nil, err
+	}
+
+	url := c.serverURL + coordApiSuffix + "create?id=" + id
+	resp, err := c.httpClient.Post(url, "application/json", bytes.NewReader(configData))
+	if err != nil {
+		return nil, fmt.Errorf("coordinator: post request failed: %w", err)
+	}
+	defer resp.Body.Close()
+
+	var res coordinatorResponse
+	err = json.NewDecoder(resp.Body).Decode(&res)
+	if err != nil {
+		return nil, fmt.Errorf("coordinator: failed to decode coordinator api response: %w", err)
+	}
+	if res.Error != "" {
+		return nil, fmt.Errorf("coordinator: coordinator api request error: %s", res.Error)
+	} else if resp.StatusCode != http.StatusCreated {
+		return nil, fmt.Errorf("coordinator: bad response status code %d", resp.StatusCode)
+	}
+
+	return &Coordinator{
+		id:     res.Id,
+		apiURL: c.serverURL + coordApiSuffix + res.Id,
+		client: c,
```

Comment:
Do we need to share whole `Client` with the coordinator? Why don't we just share `http.Client` instead? Also do they need to share same `http.Client`? This question also applies above, for the agent.

Reply:
Very good question. 

The reason behind it was that initially (and still do in a way) I wanted to keep all the request/response logic in `Client` and use that from the higher level objects (`Agent` and `Coordinator`). 

In the end I've implemented those methods (`apiGet`/`apiPost`/`apiRequest`) at the upper level (and essentially repeated them for each type) to keep the code simpler and avoid having to possibly deal with `interface{}`. 

I still think that an attempt to unify the response objects and methods should be made in the near future but opted for not doing it now mainly cause of time prioritization reasons.

> Also do they need to share same http.Client?

I believe that sharing the same `http.Client` is usually encouraged:


> https://golang.org/pkg/net/http/
> Clients and Transports are safe for concurrent use by multiple goroutines and for efficiency should only be created once and re-used.


 Also, I've provided a way to pass a custom client in case the user of the API needs more control.

Reply:
I see, let's see how it develops then.